File: python/tidegraph/source_domain.py
"""Logical incoming source slots, independent of physical wire/port slots."""
from dataclasses import dataclass


@dataclass(frozen=True)
class SourceDomain:
    edge_target: tuple[int, ...]
    input: tuple[int, ...]

    def __post_init__(self):
        for name in ("edge_target", "input"):
            values = tuple(getattr(self, name))
            if any(type(x) is not int or not 0 <= x < 2**63 for x in values):
                raise ValueError("source domain requires nonnegative int64 slots")
            object.__setattr__(self, name, values)

    def counts(self, graph):
        if len(self.edge_target) != len(graph.edges) or len(self.input) != len(graph.inputs):
            raise ValueError("source domain size mismatch")
        rows = [set() for _ in graph.nodes]
        for edge, slot in zip(graph.edges, self.edge_target):
            rows[edge.target].add(slot)
        for node, slot in zip(graph.inputs, self.input):
            rows[node].add(slot)
        for row in rows:
            if row and (min(row) != 0 or max(row) != len(row) - 1):
                raise ValueError("source domain must be dense for each node")
        return tuple(map(len, rows))

File: cpp/src/source_domain.cpp
#include "tide/source_domain.h"
#include <set>
#include <stdexcept>

namespace tide {
void compile_source_domain(Graph& g) {
  if (!g.source_domain) g.source_domain = SourceDomain{g.layout->edge_target, g.layout->input};
  const auto& domain = *g.source_domain;
  if (domain.edge_target.size() != g.edges.size() || domain.input.size() != g.inputs.size())
    throw std::invalid_argument("source domain size mismatch");
  std::vector<std::set<Index>> rows(g.nodes.size());
  for (size_t e = 0; e < g.edges.size(); ++e) rows[g.edges[e].target].insert(domain.edge_target[e]);
  for (size_t p = 0; p < g.inputs.size(); ++p) rows[g.inputs[p]].insert(domain.input[p]);
  g.source_counts.clear();
  for (const auto& row : rows) {
    if (!row.empty() && (*row.begin() != 0 || *row.rbegin() != static_cast<Index>(row.size()) - 1))
      throw std::invalid_argument("source domain must be dense nonnegative slots for each node");
    g.source_counts.push_back(row.size());
  }
}
}  // namespace tide
